### PR TITLE
Multithread string search

### PR DESCRIPTION
**Pull request by wargio**
**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

The following changes to the code has been applied:
- Added threads to the string search for a faster loading time (`izz` on 600 Mb file uses less ram and takes with 4-threads `0m28.528s` compared to single-thread `0m42.386s`).
- Fixed usage of `bin.maxstrbuf` with a proper warning message.
- Removed `HtUP *strings_db;` from `RzBinObject` since it is used only when `rz_bin_file_strings` is called and nowhere else.
- Refactored `search_str.c` and `bfile.c` for thread-safety and correctness.

### CODE REVIEW

File: librz/bin/bfile_string.c [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+	RzStrEnc encoding;
+} SearchThreadData;
+
+static bool is_data_section(RzBinFile *a, RzBinSection *s) {
+	if (s->has_strings || s->is_data) {
+		return true;
+	} else if (!s->name) {
+		return false;
+	}
+	// Rust
```
Comment (XVilka) [inline librz/bin/bfile_string.c]: A bit more elaborate comment would be nice

File: librz/bin/bfile_string.c [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+		.check_ascii_freq = std->bf->rbin->strseach_check_ascii_freq,
+	};
+
+	ut8 *buf = calloc(size, 1);
+	if (!buf) {
+		RZ_LOG_ERROR("bin_file_strings: cannot allocate string seac buffer.\n");
+		return NULL;
+	}
+
+	// beg critical section
```
Comment (XVilka) [inline librz/bin/bfile_string.c]: "beg"?

File: librz/bin/bfile_string.c [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+				break;
+			} else if (!bf->o) {
+				continue;
+			}
+
+			// find virt address.
+			bstr->paddr += bf->o->boffset;
+			bstr->vaddr = rz_bin_object_p2v(bf->o, bstr->paddr);
+
+			// beg critical section
```
Comment (XVilka) [inline librz/bin/bfile_string.c]: Same

File: librz/bin/bfile_string.c [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+
+static void bin_file_string_search_free(SearchThreadData *std) {
+	if (!std) {
+		return;
+	}
+	rz_list_free(std->results);
+	free(std);
+}
+
+static bool create_string_search_thread(RzThreadPool *pool, RzThreadLock *lock, RzBinFile *bf, size_t min_length, RzThreadQueue *intervals, HtUP *strings_db) {
```
Comment (XVilka) [inline librz/bin/bfile_string.c]: Maybe split this long line?

File: librz/bin/bfile_string.c [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+
+static int string_compare_sort(const RzBinString *a, const RzBinString *b) {
+	if (b->paddr == a->paddr) {
+		if (b->vaddr == a->vaddr) {
+			return 0;
+		} else if (b->vaddr > a->vaddr) {
+			return -1;
+		}
+		return 1;
+	} else if (b->paddr > a->paddr) {
```
Comment (XVilka) [inline librz/bin/bfile_string.c]: Maybe reverse the condition to exit quickly?

File: librz/bin/bfile_string.c [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+		return -1;
+	}
+	return 1;
+}
+
+static void string_scan_range_cfstring(RzBinFile *bf, HtUP *strings_db, RzList *results, const RzBinSection *section) {
+	// load objc/swift strings from cfstring table section
```
Comment (XVilka) [inline librz/bin/bfile_string.c]: `CFstring` I think?

File: librz/bin/bfile_string.c [hunk 7]
```diff
@@ -0,0 +1,458 @@
...
+	RzBinSection *section = NULL;
+	RzBinObject *o = bf->o;
+	if (!o) {
+		return;
+	}
+	rz_list_foreach (o->sections, iter, section) {
+		if (!section->name || section->paddr >= bf->size) {
+			continue;
+		} else if (max_interval && section->size > max_interval) {
+			RZ_LOG_WARN("bin_file_strings: search interval size (0x%" PFMT64x ") exeeds bin.maxstrbuf (0x%" PFMT64x "), skipping it.\n", section->size, max_interval);
```
Comment (XVilka) [inline librz/bin/bfile_string.c]: Split this line please

File: librz/bin/bfile_string.c [hunk 8]
```diff
@@ -0,0 +1,458 @@
...
+/**
+ * \brief  Generates a RzList struct containing RzBinString from a given RzBinFile
+ *
+ * \param  bf           The RzBinFile to use for searching for strings
+ * \param  min_length   The string minimum length
+ * \param  raw_strings  When set to false, it will search for strings only in the data section
+ *
+ * \return On success returns RzList pointer, otherwise NULL
+ */
+RZ_API RZ_OWN RzList *rz_bin_file_strings(RZ_NONNULL RzBinFile *bf, size_t min_length, bool raw_strings) {
```
Comment (XVilka) [inline librz/bin/bfile_string.c:304]: Can be `min_length == 0`? Should check with the assert and add a comment please

File: librz/bin/bfile_string.c [hunk 9]
```diff
@@ -0,0 +1,458 @@
...
+	strings_db = ht_up_new0();
+	if (!strings_db) {
+		RZ_LOG_ERROR("bin_file_strings: cannot allocate string map.\n");
+		goto fail;
+	}
+
+	if (raw_strings) {
+		// returns all the strings found on the RzBinFile
+		ut64 section_size = bf->size / pool_size;
+		if (section_size & (0x10000 - 1)) {
```
Comment (XVilka) [inline librz/bin/bfile_string.c]: extract this constant into the define and use it

File: librz/bin/bfile_string.c [hunk 10]
```diff
@@ -0,0 +1,458 @@
...
+			section_size += 0x10000;
+			section_size &= ~(0x10000 - 1);
+		}
+		if (!section_size) {
+			section_size += 0x10000;
+		}
+
+		if (max_interval && section_size > max_interval) {
+			RZ_LOG_WARN("bin_file_strings: search interval size (0x%" PFMT64x ") exeeds bin.maxstrbuf (0x%" PFMT64x "), skipping it.\n", section_size, max_interval);
```
Comment (XVilka) [inline librz/bin/bfile_string.c]: Split the long line

File: librz/bin/bfile_string.c [hunk 11]
```diff
@@ -0,0 +1,458 @@
...
+	} else if (bf->o && !rz_list_empty(bf->o->sections)) {
+		// returns only the strings found on the RzBinFile but within the data section
+		RzListIter *iter = NULL;
+		RzBinSection *section = NULL;
+		RzBinObject *o = bf->o;
+		rz_list_foreach (o->sections, iter, section) {
+			if (section->paddr >= bf->size) {
+				continue;
+			} else if (max_interval && section->size > max_interval) {
+				RZ_LOG_WARN("bin_file_strings: search interval size (0x%" PFMT64x ") exeeds bin.maxstrbuf (0x%" PFMT64x "), skipping it.\n", section->size, max_interval);
```
Comment (XVilka) [inline librz/bin/bfile_string.c]: Split

File: librz/bin/bfile_string.c [hunk 12]
```diff
@@ -0,0 +1,458 @@
...
+				continue;
+			}
+
+			if (is_data_section(bf, section)) {
```
Comment (XVilka) [inline librz/bin/bfile_string.c]: You can do `if !is_data_section() { continue }`

**Review by XVilka — CHANGES_REQUESTED** [review]
Looks great! And definitely faster

File: librz/bin/bfile_string.c (hunk 1 above)
Comment (wargio) [reply librz/bin/bfile_string.c]: that is actually a copy and paste

**Review by XVilka — APPROVED** [review]

**Pull request merged**